File: leetcode/300-400/310. 最小高度树.cpp
#include <bits/stdc++.h>
using namespace std;
vector<vector<int>> DIRS = {{1, 1}, {1, -1}, {-1, -1}, {-1, 1}};
constexpr int MOD = int(1e9 + 7);

class Solution {
public:
    // 使用邻接表存储数
    vector<vector<int>> g;

    // p1 存储的是d1对应的路径
    /**
    d1 往下走的最长路径
    d2 往下走的次长路径（如果d1[i] == d2[i], 表示i点有两条最长的路径)
    up 网上走的路径
    **/
    vector<int> d1, d2, p1, up;

    void dfs1(int u, int father) {
        /**
        计算d1, d2, p1
        **/
        d1[u] = d2[u] = 0;
        for (int j : g[u]) {
            // 遍历所有的子节点
            // 需要跳过父节点，避免反向遍历
            if (j == father) continue;
            dfs1(j, u);
            // 子路径加1（此处的边权是1）
            int d = d1[j] + 1;
            if (d >= d1[u]) {
                // 先将当前最长路径给次长路径
                d2[u] = d1[u];
                // 更新最长路径
                d1[u] = d;
                p1[u] = j;
            } else if (d >= d2[u]) {
                // 只比次大值大，更新次大值
                d2[u] = d;
            }
        }
    }

    void dfs2(int u, int father) {
        for (int j : g[u]) {
            if (j == father) continue;
            if (j == p1[u]) {
                // j在u往下走的最长路径上, 比较父节点的up和往下走的次大值
                up[j] = max(up[u], d2[u]) + 1;
            } else {
                up[j] = max(up[u], d1[u]) + 1;
            }
            dfs2(j, u);
        }
    }

    vector<int> findMinHeightTrees(int n, vector<vector<int>> &edges) {
        // 初始化g, 修改为指定大小
        g.resize(n);
        d1 = d2 = p1 = up = vector<int>(n);

        // 存储图
        for (auto &e : edges) {
            int a = e[0], b = e[1];
            g[a].push_back(b);
            g[b].push_back(a);
        }
        vector<int> res;
        dfs1(0, -1);
        dfs2(0, -1);

        int min_heigh = n + 1;
        for (int i = 0; i < n; ++i) {
            // 遍历所有节点，找到最小的高度
            min_heigh = min(min_heigh, max(d1[i], up[i]));
        }
        for (int i = 0; i < n; ++i) {
            // 再次遍历所有节点，如果该点高度等于最小高度则加入答案， 这一步也可以合并到上一步中，为了清晰则不单独写。
            if (max(d1[i], up[i]) == min_heigh) res.push_back(i);
        }

        return res;
    }
};

File: leetcode/300-400/310. 最小高度树.py
from collections import *
from functools import *
from math import *
from itertools import *
from heapq import *
from bisect import *
from typing import *
from queue import *
from string import *

DIRS = [(1, 1), (1, -1), (-1, -1), (-1, 1)]
Max = lambda a, b: b if b > a else a
Min = lambda a, b: b if b < a else a
INF = float("inf")
MOD = int(1e9 + 7)


from typing import List, Tuple
from math import inf


class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        if n == 1:
            return [0]

        # 构建无向图
        g = [[] for _ in range(n)]
        for x, y in edges:
            g[x].append(y)
            g[y].append(x)

        # max_2[i] 表示从节点 i 往下走的前两个最长路径和其来源节点
        max_2 = [[(0, -1), (0, -1)] for _ in range(n)]

        # 第一次 DFS：从下往上计算每个节点的前两大子树路径长度
        def dfs1(i: int, fa: int) -> Tuple[int, int]:
            res = [(0, -1), (0, -1)]
            for v in g[i]:
                if v == fa:
                    continue
                (child_max_len, _) = dfs1(v, i)
                current_len = child_max_len + 1
                if current_len > res[0][0]:
                    res[1] = res[0]
                    res[0] = (current_len, v)
                elif current_len > res[1][0]:
                    res[1] = (current_len, v)
            max_2[i] = res[:]
            return res[0]  # 返回最长路径

        dfs1(0, -1)

        # 第二次 DFS：换根，从父亲传来的路径往子节点传播
        ans_len = inf
        ans = []

        def dfs2(i: int, fa: int):
            nonlocal ans_len, ans

            mx1, from1 = max_2[i][0]
            mx2, from2 = max_2[i][1]

            # 更新答案
            if mx1 < ans_len:
                ans_len = mx1
                ans = [i]
            elif mx1 == ans_len:
                ans.append(i)

            # 向子节点传递信息（换根）
            for v in g[i]:
                if v == fa:
                    continue

                # 如果 v 是从 max_2[i][0] 这条最长路径来的
                if v == from1:
                    propagated = mx2 + 1
                else:
                    propagated = mx1 + 1

                # 尝试更新 max_2[v] 来自父亲的路径
                if propagated > max_2[v][0][0]:
                    max_2[v][1], max_2[v][0] = max_2[v][0], (propagated, i)
                elif propagated > max_2[v][1][0]:
                    max_2[v][1] = (propagated, i)

                dfs2(v, i)

        dfs2(0, -1)
        return ans


class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        if n == 1:
            return [0]
        g = [[] for _ in range(n)]
        degree = [0] * n
        for a, b in edges:
            g[a].append(b)
            g[b].append(a)
            degree[a] += 1
            degree[b] += 1
        q = deque(i for i in range(n) if degree[i] == 1)
        ans = []
        while q:
            ans.clear()
            for _ in range(len(q)):
                a = q.popleft()
                ans.append(a)
                for b in g[a]:
                    degree[b] -= 1
                    if degree[b] == 1:
                        q.append(b)
        return ans


class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        if n == 1:
            return [0]

        g = [[] for _ in range(n)]
        for x, y in edges:
            g[x].append(y)
            g[y].append(x)
        parents = [0] * n

        def bfs(start: int):
            vis = [False] * n
            vis[start] = True
            q = deque([start])
            while q:
                x = q.popleft()
                for y in g[x]:
                    if not vis[y]:
                        vis[y] = True
                        parents[y] = x
                        q.append(y)
            return x

        x = bfs(0)  # 找到与节点 0 最远的节点 x
        y = bfs(x)  # 找到与节点 x 最远的节点 y

        path = []
        parents[x] = -1
        while y != -1:
            path.append(y)
            y = parents[y]
        m = len(path)
        return [path[m // 2]] if m % 2 else [path[m // 2 - 1], path[m // 2]]
